### ISSUE TITLE
save and restore is working or not ?

### ISSUE THREAD

**Issue by yichunsung** [rstudio/tensorflow#197]
Hi, I try to use Tensorflow mnist data to test ` tf$train$Saver$save()` and `restore`, but I can not restore my variable.

Same code I use Python to test, it is working well. Therefore, I think it might be problem for R ?

This is  model saver file:
```{r}

hiddenLayer_1_Weight <- tf$Variable(tf$random_normal(shape(784L, 400L)), name = "h1W1")
hiddenLayer_1_bias <- tf$Variable(tf$zeros(shape(1, 400L)), name = "h1b1")

hiddenLayer_2_Weight <- tf$Variable(tf$random_normal(shape(400L, 250L)), name = "h2W2")
hiddenLayer_2_bias <- tf$Variable(tf$zeros(shape(1, 250L)), name = "h2b2")

hiddenLayer_3_Weight <- tf$Variable(tf$random_normal(shape(250L, 40L)), name = "h3W3")
hiddenLayer_3_bias <- tf$Variable(tf$zeros(shape(1, 40L)), name = "h3b3")

outputLayer_Weight <- tf$Variable(tf$random_normal(shape(40L, 10L)), name = "opW")
outputLayer_bias <- tf$Variable(tf$zeros(shape(1, 10L)), name = "opb")


##  do something for model building.......
.......
................
......................
## Setting session and saver
sess <- tf$Session()
saver <- tf$train$Saver()
init <- tf$global_variables_initializer()
sess$run(init)

##  do something for model training.......

# Save
print("==== Saver is start ====")
saver$save(sess, '/mydir/DNN_model_MNIST', write_meta_graph = FALSE)
print("Save successfully")
print("==== Saver is end ====")
```

> # After training, I closed Rstudio and reopened it, and run my restore script. However, when I computed the accuracy for the model, it's not a good result. The result looks like no training.

```{r}
hiddenLayer_1_Weight <- tf$get_variable("h1W1", shape(784L, 400L))
hiddenLayer_1_bias <- tf$get_variable("h1b1", shape(1, 400L))

hiddenLayer_2_Weight <- tf$get_variable("h2W2", shape(400L, 250L))
hiddenLayer_2_bias <- tf$get_variable("h2b2", shape(1, 250L))

hiddenLayer_2_Weight <- tf$get_variable("h3W3", shape(400L, 250L))
hiddenLayer_3_bias <- tf$get_variable("h3b3", shape(1, 40L))

outputLayer_Weight <- tf$get_variable("opW", shape(40L, 10L))
outputLayer_bias <- tf$get_variable("opb", shape(1, 10L))

## Setting session and saver
sess <- tf$Session()
init <- tf$global_variables_initializer()
sess$run(init)
saver <- tf$train$Saver()
print("Restore ")
saver$restore(sess, '/mydir/DNN_model_MNIST')
print("Restore successfully")

```

**Comment by yichunsung** [comment]
It's interesting that I use Python to restore the Variable build from R. I still got bad accuracy.
However, I use R to restore the Variable build from Python. It's working and give me a good accuracy.

**Comment by jjallaire** [comment]
Could you provide a complete (and minimal) example that exhibits this problem? (that will make it much easier to investigate on our end).

**Comment by yichunsung** [comment]
Hi, It's my simple example for training model and saver.
After 1000 times for training, I got **0.8224** for accuracy.
However, When I closed Rstudio, and reopen it next time and restore the saver to test it.
It's only **0.0638** for accuracy.

P.S. Same code in python is working well.

## **It's training code:**
```{r}
### Sample for saver
library(tensorflow)
library(magrittr)
### MNIST for TensorFlow
### Data loading
datasets <- tf$contrib$learn$datasets
mnist <- datasets$mnist$read_data_sets("MNIST-data", one_hot = TRUE)

### Placeholder 
xs <- tf$placeholder(tf$float32, shape(NULL , 784L))
ys <-tf$placeholder(tf$float32, shape(NULL, 10L))

### Model Building 
### Hidden layer 
hiddenLayer_Weight <- tf$Variable(tf$random_normal(shape(784L, 40L)), name = "h1W1")
hiddenLayer_bias <- tf$Variable(tf$zeros(shape(1, 40L)), name = "h1b1")
hiddenLayer_Wx_plus_b <- tf$matmul(xs, hiddenLayer_Weight)+ hiddenLayer_bias
hiddenLayer_output <- tf$nn$sigmoid(hiddenLayer_Wx_plus_b)

### output layer 
outputLayer_Weight <- tf$Variable(tf$random_normal(shape(40L, 10L)), name = "opW")
outputLayer_bias <- tf$Variable(tf$zeros(shape(1, 10L)), name = "opb")
outputLayer_Wx_plus_b <- tf$matmul(hiddenLayer_output, outputLayer_Weight)+ outputLayer_bias
output_layer_result <- tf$nn$softmax(outputLayer_Wx_plus_b)

### loss function, Gradient descent and learning rate
cross_entropy <- tf$reduce_mean(-tf$reduce_sum(ys*tf$log(output_layer_result), 
                                               reduction_indices = 1L))
learning_rate <- 0.4
train_step <- tf$train$AdamOptimizer(learning_rate)$minimize(cross_entropy)

### compute accuracy function
compute_accuracy <- function(model_result, v_xs, v_ys){
  y_pre <- sess$run(model_result, feed_dict = dict(xs = v_xs))
  correct_prediction <- tf$equal(tf$argmax(y_pre, 1L), tf$argmax(v_ys, 1L))
  accuracy <- tf$cast(correct_prediction, tf$float32) %>% tf$reduce_mean(.)
  result <- sess$run(accuracy, feed_dict = dict(xs = v_xs, ys = v_ys))
  return(result)
}

### Training
### Setting session and saver
sess <- tf$Session()
saver <- tf$train$Saver()
init <- tf$global_variables_initializer()
sess$run(init)

### Running
for (i in 1:1000){
  batches <- mnist$train$next_batch(100L)
  batch_xs <- batches[[1]]
  batch_ys <- batches[[2]]
  sess$run(train_step, feed_dict = dict(xs = batch_xs, ys = batch_ys))
  if(i %% 50 == 0){
    print(paste("Step =>", i, sep = " "))
    print(compute_accuracy(output_layer_result, mnist$test$images, mnist$test$labels))
  }
}
print("==== Saver is start ====")
### Step = 1000, compute_accuracy = 0.8224
saver$save(sess, '/Github_rep/R-Tensorflow_Notebook/DNN/model_saver/issue_test/Saver_test', write_meta_graph = FALSE)
print("Save successfully")
print("==== Saver is end ====")

```
## **It's restore code:**
```{r}
### sample for restore
library(tensorflow)
library(magrittr)
### MNIST for TensorFlow
#### Data loading
datasets <- tf$contrib$learn$datasets
mnist <- datasets$mnist$read_data_sets("MNIST-data", one_hot = TRUE)
### Restore Weight and bias
hiddenLayer_Weight <- tf$get_variable("h1W1", shape(784L, 40L))
hiddenLayer_bias <- tf$get_variable("h1b1", shape(1, 40L))
outputLayer_Weight <- tf$get_variable("opW", shape(40L, 10L))
outputLayer_bias <- tf$get_variable("opb", shape(1, 10L))
### Placeholder 
xs <- tf$placeholder(tf$float32, shape(NULL , 784L))
ys <-tf$placeholder(tf$float32, shape(NULL, 10L))
### Structure of model
hiddenLayer_Wx_plus_b <- tf$matmul(xs, hiddenLayer_Weight)+ hiddenLayer_bias
hiddenLayer_output <- tf$nn$sigmoid(hiddenLayer_Wx_plus_b)

outputLayer_Wx_plus_b <- tf$matmul(hiddenLayer_output, outputLayer_Weight)+ outputLayer_bias
output_layer_result <- tf$nn$softmax(outputLayer_Wx_plus_b)

### compute accuracy function
compute_accuracy <- function(model_result, v_xs, v_ys){
  y_pre <- sess$run(model_result, feed_dict = dict(xs = v_xs))
  correct_prediction <- tf$equal(tf$argmax(y_pre, 1L), tf$argmax(v_ys, 1L))
  accuracy <- tf$cast(correct_prediction, tf$float32) %>% tf$reduce_mean(.)
  result <- sess$run(accuracy, feed_dict = dict(xs = v_xs, ys = v_ys))
  return(result)
}

### Setting session and restore
sess <- tf$Session()
init <- tf$global_variables_initializer()
sess$run(init)
saver <- tf$train$Saver()
saver$restore(sess, '/Github_rep/R-Tensorflow_Notebook/DNN/model_saver/issue_test/Saver_test')
print("Restore successfully")
print("===Test accuracy===")
print(compute_accuracy(output_layer_result, mnist$test$images, mnist$test$labels)) 
## only 0.0638 for accuracy ..........?


```

**Comment by javierluraschi** [comment]
@yichunsung  While running the example above, I got the same values (`0.7893`) for training and after loading using the `tensorflow` devel version.

Which is your TensorFlow session? Can you share `sessionInfo()`? Could you try the devel version of tensorflow by running:

```r
devtools::install_github("rstudio/tensorflow")
```

and retrying?

**Comment by yichunsung** [comment]
@javierluraschi , in initially, I used  TensorFlow 1.3.1 and R 3.4.1.
I follow your suggestion, and updated the package of TensorFlow to 1.4 and R 3.4.2.
After updated version, It's seemed that I can get the same value for training and restore.
However, I tried it for 5 times after new version Tensorflow and R, I trained, saved,  restore, and delete my save file from my disk, and do the same processes again, it's weird that I can't get same value in second times, but can successfully get same value in others times.

Do you think devel version is better than 1.4 ?
